crates/wysiwyg/src/dom/find_extended_range.rs

```diff
@@ -0,0 +1,221 @@
+        find_range::find_range(self, s, e)
+    }
+
+    /// Returns a selection from given start and end, extended up to the next
+    /// structure node, list item or linebreak on each side.
+    pub fn find_extended_selection(
+        &self,
+        start: usize,
+        end: usize,
+    ) -> (usize, usize) {
+        let range = find_range::find_range(self, start, end);
+        let leaves: Vec<&DomLocation> = range
+            .leaves()
+            .filter(|loc| loc.position < range.end())
+            .collect();
+        if leaves.is_empty() {
+            return (start, end);
+        }
+
+        let first_leaf = leaves.first().unwrap();
+        let last_leaf = leaves.last().unwrap();
+        let extended_start = start - self.extended_offset_before(first_leaf);
+        let extended_end = end + self.extended_offset_after(last_leaf);
+
+        (extended_start, extended_end)
+    }
+
+    /// Returns the offset before a location until a structure
+    /// node, list item or linebreak is found.
+    fn extended_offset_before(&self, location: &DomLocation) -> usize {
+        if location.kind == DomNodeKind::LineBreak {
+            // Item at location is a linebreak, no need to iterate
+            return location.start_offset;
+        }
+
+        let mut iter = self.iter_from_handle(&location.node_handle).rev();
+        // Remove location and use offset instead
+        iter.next();
+        let mut offset = location.start_offset;
+
+        for node in iter {
+            if node.is_text_node() {
+                offset += node.text_len();
+            } else if node.is_line_break() {
+                offset += node.text_len();
+                break;
+            } else if node.is_block_node() || node.is_list_item() {
+                break;
+            }
+        }
+
+        offset
+    }
+
+    /// Returns the offset after a location until a structure
+    /// node, list item or linebreak is found.
+    fn extended_offset_after(&self, location: &DomLocation) -> usize {
+        if location.kind == DomNodeKind::LineBreak
+            || self.is_last_child_of_list(&location.node_handle)
+        {
+            // Item at location is a linebreak or the last child of a list, no need to iterate
+            return location.length - location.end_offset;
+        }
+
+        let mut iter = self.iter_from_handle(&location.node_handle);
+        // Remove location and use offset instead
+        iter.next();
+        let mut offset = location.length - location.end_offset;
+
+        for node in iter {
+            if node.is_text_node() {
+                offset += node.text_len();
+                if self.is_last_child_of_list(&node.handle()) {
+                    break;
+                }
+            } else if node.is_line_break() {
+                offset += node.text_len();
+                break;
+            } else if node.is_block_node() || node.is_list_item() {
+                break;
+            }
+        }
```

Comment:
Similar to above, I think legibility is improved by separating the offset increment and the break logic